src/com/facebook/buck/intellij/ideabuck/src/com/facebook/buck/intellij/ideabuck/actions/BuckRunSelectedTestAction.java

```diff
@@ -0,0 +1,180 @@
+      String testTarget = getTestTarget(buckFile, relative);
+      String name = "";
+      String testSelectors = "";
+      if (method != null) {
+        name = psiClass.getName() + "#" + method.getName();
+        testSelectors = psiClass.getName() + "#" + method.getName();
+      } else {
+        name = element.getContainingFile().getName();
+        if (element instanceof PsiClass) {
+          testSelectors = psiClass.getQualifiedName();
+        }
+      }
+      RunManagerImpl runManager = (RunManagerImpl) RunManager.getInstance(project);
+      TestConfigurationType type = new TestConfigurationType();
+
+      RunnerAndConfigurationSettingsImpl runnerAndConfigurationSettings = (RunnerAndConfigurationSettingsImpl)runManager.createRunConfiguration(name, type.getConfigurationFactories()[0]);
+      TestConfiguration testConfiguration = (TestConfiguration) runnerAndConfigurationSettings.getConfiguration();
+      testConfiguration.data.target = testTarget;
+      testConfiguration.data.testSelectors = testSelectors;
+      runManager.addConfiguration(runnerAndConfigurationSettings, false);
+      runManager.setSelectedConfiguration(runnerAndConfigurationSettings);
+
+      Executor executor = ExecutorRegistry.getInstance().getRegisteredExecutors()[0];
+      ProgramRunnerUtil.executeConfiguration(project, runnerAndConfigurationSettings, executor);
+    }
+  }
+
+  @Override
+  public void update(final AnActionEvent event) {
+    super.update(event);
+
+    PsiElement element = event.getData(LangDataKeys.PSI_ELEMENT);
+    final Presentation presentation = event.getPresentation();
+
+    if (element instanceof PsiMethod) {
+      PsiMethod method = (PsiMethod) element;
+      method.getModifierList().getAnnotations();
+      PsiAnnotation[] annotations =  method.getModifierList().getAnnotations();
+      boolean isTestMethod = false;
+      isTestMethod = isTestMethod(method);
+      if (isTestMethod) {
+        presentation.setEnabled(true);
+      } else {
+        presentation.setEnabled(false);
+      }
+    } else if (element instanceof PsiClass) {
+      PsiClass psiClass = (PsiClass) element;
+      boolean isTestClass = false;
+      for (PsiMethod method : psiClass.getAllMethods()) {
+        isTestClass = isTestClass || isTestMethod(method);
+      }
+      if (isTestClass) {
+        presentation.setEnabled(true);
+      } else {
+        presentation.setEnabled(false);
+      }
+    }
+    else {
+      presentation.setEnabled(false);
+    }
+  }
+
+  /**
+   * Get a test target for a given buck file.
+   *
+   * @param buckFile a {@link VirtualFile} which refers to the buckfile.
+   * @param buckPath an {@link String} which denotes the path to the Buck file.
+   *
+   * @return a {@link String} representing a test target to invoke.
+   */
+  @NotNull
+  private String getTestTarget(VirtualFile buckFile, String buckPath) {
```

Comment:
I thought that was how VirtualFile buckFile = BuckFileUtil.getBuckFile(virtualFile); Worked?